### ISSUE TITLE
Akaze's implementation seems to have issues with rotational invariance

### ISSUE THREAD

**Issue by stephanemagnenat** [rust-cv/cv#63]
I am trying to use Akaze for an application that searches correspondences between a template and an image seen by a camera.

As a start, I have written the common pipeline building on the excellent tutorial of Rust CV. I am using Akaze with threshold 0.001, looking for two nearest neighbors, and using a distance ratio criteria of 0.8 between the first best match and the second one (best_dist < 0.8 * next_best_dist), as suggested in the [Akaze OpenCV tutorial](https://docs.opencv.org/4.x/db/d70/tutorial_akaze_matching.html). Besides that, I am using the functions `matching` and `symmetric_matching` from the tutorial as is.

For comparison, I created a similar pipeline in Python using OpenCV (4.7.0.72). When only scaling is involved, both Rust CV's Akaze and Open CV Akaze perform well:

*Open CV, before matching: left: 2514 keypoints, right: 820 keypoints*:
![correspondences_opencv_scale](https://github.com/rust-cv/cv/assets/390278/d3fbe72d-39a0-4144-9a90-0e5a015b105a)

*Rust CV, before matching: left: 5353 keypoints, right: 1480*:
![correspondences_rustcv_scale](https://github.com/rust-cv/cv/assets/390278/598bc5ee-3482-45ec-ae68-e655ed786dc8)

However, when I rotate the small image, Rust CV completely fails while Open CV works like a charm:

*Open CV, before matching: left: 2514 keypoints, right: 1207 keypoints*:
![correspondences_opencv_rot](https://github.com/rust-cv/cv/assets/390278/ca212f0e-7db4-43b0-9f1f-8491fbb815ea)

*Rust CV, before matching: left 5353 keypoints, right: 2490*
![correspondences_rustcv_rot](https://github.com/rust-cv/cv/assets/390278/659220b9-7b11-4bc8-9ff7-ab096022014e)

So, it looks to me that there is something very broken with the rotational invariance of RustCV's Akaze implementation.
Was this implementation ever validated against the OpenCV one?

I am also surprised by the different numbers of matches, as I would expect more similar ones between the two implementations.

**Comment by stephanemagnenat** [comment]
I believe I have found some fishy code: `angs[idx] = res_y[idx].atan2(res_y[idx]);` ([here](https://github.com/rust-cv/cv/blob/2e250a6968ce307e9385ad00584e1585f02b466b/akaze/src/scale_space_extrema.rs#L237)). It looks like it should be something like `res_y[idx].atan2(res_x[idx])` instead.

**Comment by xd009642** [comment]
So the implementation was based off an existing Rust implementation and I think another independent implementation instead of OpenCV. I'm not sure if it was compared to OpenCVs implementation - but for more complicated algorithms I personally avoid that sometimes as OpenCV applies a lot of heuristics that may not be in the paper (from experience the OpenCV HOG adds in image pyramids and other things other implementations and the paper don't use). 

The fishy line you spotted does seem to be a problem so a PR for that would be good. Does it improve the output at all?

I'm also not super familiar with this area - only mildly familiar as a casual user - so am likely not best to comment but I can try to help out. One question would be how does the matcher impact this? 

I assume the visualisation is just showing the points that match between both and not the points where no match was found. Is there some worth in showing all the points identified in the rotated and non-rotated versions including unmatched ones to see if something's up. Although there are _a lot_ of points so not sure if that's useful at all :thinking:

**Comment by stephanemagnenat** [comment]
Yes showing all points is likely to be a lot. I think that showing only matching lines, as now, is enough so far.

I found other bugs in the translation from the [original C++ implementation](https://github.com/pablofdezalc/akaze) the previous Rust implementation is based on. You can follow my progress in [this PR](https://github.com/rust-cv/cv/pull/64). I'm comparing the Rust code with that original implementation.

The one line correction above doesn't improve the performance so far.

**Comment by stephanemagnenat** [comment]
After comparing the evolutions of the same image, I can confirm that they are very different, so probably there is a bug already at that level in the Rust version.

**Comment by stephanemagnenat** [comment]
Already `rust_evolution_0_Lt.exr` looks suspicious. It seems that the Gaussian filter didn't work properly.

**Comment by xd009642** [comment]
I guess it's an implementation detail of the gaussian blur regarding padding and the imageproc one applies some form of padding whereas cv doesn't

**Comment by stephanemagnenat** [comment]
No, it is worst than that. The Gaussian blur is seriously degraded. I have the impression that there is an incorrect optimization. But there are so many other bugs that at the moment I'll leave with `gaussian_blur_f32` that gives the same result as OpenCV after [I fixed it](https://github.com/image-rs/imageproc/pull/528).

Fyi my approach is to dump the intermediate generated images (e.g. for the different maps of the evolutions) and compare with the ones from the original C++ implementation, that I have instrumented to generate the corresponding images at the same positions. That shows that, for example, the Scharr filters are also incorrect.

**Comment by xd009642** [comment]
Okay, I know this is used as a key part of the VSLAM toolbox for reconstructing 3D point clouds from things like KITTI. Once the PR is ready it would be good to rerun that and make sure it still works as expected (oh god I might have to get vulkan working on my linux machine). I may also prod @vadixidav to take a look at it as he's more familiar with the code/domain than I am.

I'm aware that code like this can be tricky to test but it would be good to see some tests to catch certain issues discovered and try and prevent any regressions. I'll try and allocate some time later this week or next week to look into the gaussian blur some more

**Comment by stephanemagnenat** [comment]
Ok, good! I fully agree that it is worth checking that there is no regression.

Also, there is quite some re-implementation of features present in imageproc (e.g. Gaussian filter, Schaar filters, etc.). Personally I would prefer using the versions there, or contributing better implementations there if needed (but maybe the current implementations there might be faster now).

**Comment by xd009642** [comment]
When this crate was started imageproc/image didn't support having floats as a pixel type and iirc there had been an open issue for quite a while and it needed a big refactor to be possible. It may be that the reimplementations are no longer necessary

**Comment by stephanemagnenat** [comment]
The diffusion function `nonlinear_diffusion::calculate_step` is not consistent with the one of the C++ reference implementation: somehow it seems that it doesn't diffuse enough.

**Comment by xd009642** [comment]
There's not currently an established best practice, and I know one of the goals was working on things like embedded linux boards which makes the intrusiveness even more painful. I think for the current PRs don't submit it.

My thoughts are that we may want to add a ability to compile with something like `RUSTFLAGS="--cfg cv_debug"` and then we could feature gate it like `#[cfg(cv_debug)]` in the code so it's not compiled into the library unless the user explicitly wants it. However, before we make such a decision would need to discuss with other maintainers/users some may have better ideas or issues with that approach.

**Comment by stephanemagnenat** [comment]
We could also gate some introspection features behind `#[cfg(test)]`. For example for Akaze, we could have some sort of  `AkazeInspector` trait that provides callbacks to be called at different points. They will not be available in release.

**Comment by xd009642** [comment]
hmm - my initial thought is that such introspection is also useful for engineers working on computer vision applications so I'd rather it was configurable by users as well. But I may be biased as someone who's had to patch opencv with that kind of debug dumping before

**Comment by stephanemagnenat** [comment]
Sure. I have them all around, but I'll refactor that into two commits before submitting the PR for inclusion.

**Comment by vadixidav** [comment]
That sounds like a good solution. Personally I also think exposing enough APIs to dig into all the intermediaries so that we can dump debug data from an external tool would be acceptable as well. For instance, running a step of diffusion manually should be possible using a lower level API. We can add this at a later time. I would be okay with having a cfg for enabling debug information for the time being.

**Comment by vadixidav** [comment]
So, it has been a while, but I recalled something in particular that I was working on to make the algorithm behave more like the original, but it had some issues. I believe I reverted it, and now it uses nearest neighbor algorithm, see here: https://github.com/rust-cv/akaze/commit/c8de15507e8d59485bea47d40b25fc796ceb3253. That is the original commit on the archived repository. See here for the current: https://github.com/rust-cv/cv/blob/main/akaze/src/image.rs#L118.

**Comment by vadixidav** [comment]
I incorporated the change I recommended above along with @stephanemagnenat 's changes in #65.

**Comment by stephanemagnenat** [comment]
Thankz @vadixidav, I also had the same function for resizing (albeit implemented differently) in my not-yet-updated changes branch on my local computer (I wanted to clean-up the debugging changes before updating the PR branch, but maybe I should keep it updated for information, so you can freely extract pieces from it). I have other changes such as correcting the diffusion step. I will test your changes, update my branch and report here.

Regarding the check you added in `mldb_fill_values`, I didn't add it because the original code doesn't do an out of bounds access where the Rust code does (I added a check in the original code to be sure), so I suspect that there are bugs elsewhere that lead to this erroneous access and that once solved, this check can be removed.

**Comment by stephanemagnenat** [comment]
One thing regarding debugging, I believe that to ensure non-regression of this algorithm, ideally we should have points to dump intermediate images of the internal space and compared to them of the original algorithm. That is how I managed to debug so far.

**Comment by stephanemagnenat** [comment]
My further changes leading to correct internal images after `create_nonlinear_scale_space` are now in #66.

There are still bugs ahead, I suspect that at least the Scharr filters implementation is broken. I continue my debugging and will update #66 with further corrections, or directly #65 if we want me to push directly there.

**Comment by stephanemagnenat** [comment]
The general Scharr filters were inverted! I fixed that, and simplified the code using imageproc's `separable_filter` function. Again, I didn't deleted the unused functions, but I think that we should, do you agree @xd009642 and @vadixidav?

**Comment by xd009642** [comment]
I'm personally fine with deleting unused functions if they're not public. If they're public takes a bit more of a think because of semver

**Comment by stephanemagnenat** [comment]
They are not public! Ok, I'll fold their deletion in the respective commits, it is done for the Scharr filters.

For the Gaussian, they are public in the module but not re-exported at the level of the library. I guess it is also ok to delete them as they are not in the public API, right?

**Comment by stephanemagnenat** [comment]
Ok, with the commits in #66 so far, the internal spaces are similar to an epsilon!

**Comment by stephanemagnenat** [comment]
The C++ implementation uses 8 comparison points for `find_scale_space_extrema` while the Rust one uses 5, with in addition a bug in a bound.

**Comment by stephanemagnenat** [comment]
Both implementations new find keypoints at similar positions. However matching is still not on par. Next step is to compare keypoints themselves.

**Comment by vadixidav** [comment]
Please delete unused functions, git will remember them if needed. Don't worry about them.

**Comment by vadixidav** [comment]
This is all very great investigation. Considering how central akaze is to everything we are doing, this is really excellent for the project as a whole. Thanks so much again for your investigation. I will look once more over the PR before we merge.

**Comment by stephanemagnenat** [comment]
Thank you for your kind words! It's really great to have Akaze in Rust, and I hope that we can make it reliable and performant (see #67), and therefore promote the use of Rust in CV, to everyone's benefit!

**Comment by stephanemagnenat** [comment]
So, there was a nasty bug in computing angles: `cv::fastAtan2` returns a value in [0:360], while `f32::atan2` returns one in [-π:π]. The degree to angle ratio was corrected but not the range. So half of the points had wrong orientation! With this bug fixed, quality is similar to the C++ implementation:

![correspondences](https://github.com/rust-cv/cv/assets/390278/8b8f655a-97b4-430e-9aeb-645e7d23f955)

@vadixidav also at least on my test image, there are no image access out of bound using the original code, which would make ce8f7cef88d946d6487d886736b12366f9630626 is not necessary. However, using the current Rust version that adds 0.5 to the coordinate before sampling the descriptor in `mldb_fill_values`, there is an out-of-bound access. I am under the impression the that 0.5 addition is actually more correct and that we should keep it as well as the check, but input is welcome!

**Comment by stephanemagnenat** [comment]
So, using #69 I have been running this version of Akaze on the  [2017 COCO validation dataset](http://images.cocodataset.org/zips/val2017.zip) (5k images), converted to grayscale. The bound check is necessary if we have the +0.5 on `xf` and `yf`. Without the 0.5 the check is not necessary.

So, two options:
1. We are convinced that the +0.5 is more correct and we keep as is.
2. We are not convinced, and we revert to the original algorithm and remove the boundary check.

My current feeling is to move this question to a further issue, by reverting to the original algorithm in the current PR and moving the +0.5 and the check to another PR for further investigation/discussion. That would allow us to move this PR quickly.

**Comment by stephanemagnenat** [comment]
I think that PR #69 is ready to be merged. @vadixidav I let you choose whether you wish to keep or remove the boundary check you introduced in ce8f7cef88d946d6487d886736b12366f9630626.

**Comment by vadixidav** [comment]
If you have confirmed the bounds check is no longer necessary then I will just run a test against my large dataset just in case and if it works fine then I will remove the check.

**Comment by stephanemagnenat** [comment]
Perfect! Can you do so and merge?

**Comment by stephanemagnenat** [comment]
@vadixidav, @xd009642 can we merge this please? It is in the way of more changes.

@vadixidav if you do not have time to run this PR on your large dataset, may I suggest to keep the check for now, we can always remove it later once you have run it.

**Comment by xd009642** [comment]
So I'm personally happy to merge this with the bounds check without the run on the large dataset - but vadixidav is the boss so I'll defer to his decision.

**Comment by xd009642** [comment]
@stephanemagnenat Checked with vadixidav on the rust-cv discord and we're happy to merge so I've just merged it in :tada:
